### PR TITLE
Enable operator operation without kubernetes.

### PR DESCRIPTION
**Pull request by pruiz**
Hi, latelly I've been working on setting up cilium with nomad, instead of kubernetes. By following some advices from @joestringer at https://github.com/cilium/cilium/issues/18334, and inspecting some documentation and guidance on howto setup cilium with docker, I've been able to integrate cilium agent, CNI & docker plugins to nomad. 

However, the only component not working out of the box is the operator (actually the generic operator). This is due to some code paths assuming kubernetes availability, and failing (obviously) when no k8s apiserver that can be contacted.

This PR addresses this issues by adding a new command argument (--disable-k8s-services, w/ false as default), which enables basic usage of the operator (lease cleanup, ipam address gc, etc.), but disables any k8s-specific code/features.

There are some more missing pieces/features which I would like to add to operator in future PRs, like being able to provide/distribute policy statements to agents from kvstore, instead of using kubernetes CRDs, etc. But this is the bare initial support for operator w/o kubernetes.

PD: For nomad users interested on this, this github issue is the actual place where using cilium w/ nomad is being handled (https://github.com/hashicorp/nomad/issues/12120)

### CODE REVIEW

File: operator/option/config.go
Comment (joestringer) [inline operator/option/config.go]: One nit, I would suggest using `enable-k8s-services` here and reversing the logic. Otherwise we regularly end up with double-negatives in the code like `!disableFoo` which contributors often find confusing.

File: operator/cmd/root.go
Comment (joestringer) [inline operator/cmd/root.go]: I'm curious if you have any thoughts on the rationale between just using `k8s.IsEnabled()` here vs. the new option. Is there a meaningful difference?

File: operator/cmd/root.go
Comment (joestringer) [inline operator/cmd/root.go:464]: Maybe another question, I see `SyncK8sServices` as a config option here. What's the difference between this and the new option?

**Review by joestringer — COMMENTED** [review]
I'm actually not sure whether to mark this as `release-note/minor` or `release-note/major`, as we could consider this to be a major announcement of initial Nomad support :tada: 

I ran the sanity checker CI jobs and I can kick off the full CI suite on demand, let me know if/when you're ready to run those.

Typically I would wait for the contributor to mark the PR as "ready for review" before providing comments, but since we've had some discussions before and you pinged me in the description, I took an early look. Couple of minor comments below.

File: operator/option/config.go
Comment (pruiz) [reply operator/option/config.go]: Sure, I'll update it to enable-k8-services, with default=true.

File: operator/cmd/root.go
Comment (pruiz) [reply operator/cmd/root.go:464]: I was trying to convey 'enable-k8-services' as a root toggle to disable all k8s-related features, otherwise the user needs to specify quite a handfull of options in order to 'completely' disable reliance on k8s being present.. The full list of arguments required would be something like: --disable-cnp-status-updates=true --enable-k8s-endpoint-slice=false --skip-crd-creation=true --synchronize-k8s-nodes=false --synchronize-k8s-services=false  --disable-endpoint-crd=true --enable-k8s-api-discovery=false.. along with the new --enable-k8s-services=false. Which seemed like really hard to guess for a newcomer.

However, operatorOption is not really available everywhere, but k8s.IsEnabled() is, and it is already being used as a main k8s toggle at other parts of the codebase (not only on operator, but also agent, etc.). That's why I coded it like, --enable-k8s-services=false will en up making k8s.IsEnabled() return false, and then use such means to conditionally enable anything else, and in some cases warn or inform the user that this or that k8s-related service was not enabled due to k8s not being available/enabled.

Comment (joestringer) [reply operator/cmd/root.go:464]: Ah I see, it makes sense to me for there to be one toggle for all k8s logic. When I read `enable-k8s-services`, I assumed that this meant only the "k8s service" (virtual IP / load balancing / SNAT / DNAT) type logic. Reading your response above, it sounds like you meant "services" in the sense that these are features that k8s provides. Could we just use a new flag something like `--enable-k8s` instead? Then it would be clear that it is indicating _all_ kubernetes support and remove the name conflict with a kubernetes-specific feature (Services).

Otherwise :+1: makes sense. We are currently working on refactoring the logic into more modular and convenient "Cells" (basically just application modules with self-contained logic). This effort should make this code simpler over time, but I don't think that your PR needs to wait for such an effort. Something close to what you have proposed already seems fine to me. I will CC @joamaki just as a heads up that the ability to disable k8s entirely is a community ask, in case it impacts that refactoring effort.

Comment (lazedo) [reply operator/cmd/root.go:464]: I vote for `--disable-k8s` (false by default)

Comment (joestringer) [reply operator/cmd/root.go:464]: We are trying to avoid double-negatives in the codebase, because they are frequently confusing to developers and users. This is why I'm suggesting `--enable-k8s` with a default to true.

Comment (pruiz) [reply operator/cmd/root.go:464]: I'll update tomorrow with --enable-k8s (w/ default=true) :)

**Comment by pruiz** [comment]
Updated branch w/ changes and rebased against current master.

File: operator/cmd/flags.go
Comment (joestringer) [inline operator/cmd/flags.go]:
```suggestion
	flags.Bool(operatorOption.EnableK8s, true, `Enables operation of kubernetes-related services/controllers`)
```

**Review by joestringer — CHANGES_REQUESTED** [review]
Minor nit. Could you also squash all the patches together?

You will also need to run `make && make -C Documentation update-cmdref` after changing the below in order to properly update the docs. Then the CI targets should pass.

**Comment by pruiz** [comment]
Fixed cli help reference, rebased and squashed changes.

**Comment by joestringer** [comment]
The tests failed because the autogenerated docs need to be updated. Please be sure to build the binaries locally and then run 'make -C Documentation update-cmdref', amend the commit to include the latest changes, and re-push the PR.

**Comment by pruiz** [comment]
Using update-cmdref fails on my local (probably due to using osx), so I've
been trying to match it by hand.. I'll try again or move the development to
a docker environment.

**Review by joestringer — APPROVED** [review]
LGTM. I'll run the full CI. It would be also good to get a review from @cilium/operator.

**Comment by joestringer** [comment]
Testing is passing, just looking for one more reviewer on the operator side. I think that the main concern this will draw over time is whether we can better structure the operator to support enabling these sorts of functionalities in a more modular manner, rather than scattering conditional checks around the code. At the same time, I'm not sure if we've got to a point where the necessary abstractions are clear so I don't have strong opinions about whether we should try to address that now as we begin these steps or whether we can just address that over time. Hopefully someone with a better familiarity with that part of the codebase could provide some thoughts in that regard.

File: operator/cmd/flags.go
Comment (christarazi) [inline operator/cmd/flags.go]:
```suggestion
	flags.Bool(operatorOption.EnableK8s, true, `Enable operation of Kubernetes-related services/controllers when using Cilium with Kubernetes`)
```

File: operator/option/config.go
Comment (christarazi) [inline operator/option/config.go]:
```suggestion
	// EnableK8s operation of Kubernetes-related services/controllers.
	// Intended for operating cilium with CNI-compatible orchestrators
	// other than Kubernetes. (default is true)
```

**Review by christarazi — COMMENTED** [review]
LGTM pending very minor nits

File: operator/cmd/flags.go
Comment (pruiz) [reply operator/cmd/flags.go]: Hi, I've pushed an updated branch with this fixes applied.

File: operator/option/config.go
Comment (pruiz) [reply operator/option/config.go]: Hi, I've pushed an updated branch with this fixes applied.

**Comment by joestringer** [comment]
Sorry, looks like something else went into the tree that caused a conflict. Can you rebase against the latest master branch?

**Review by christarazi — APPROVED** [review]

**Review by nebril — APPROVED** [review]
LGTM!

**Comment by joestringer** [comment]
This seems like it's got the necessary approvals, can you rebase against master? After that we can trigger CI and hopefully merge.

**Comment by pruiz** [comment]
@joestringer  I'll try, but this is getting a bit too much overhead to get merged. I understand each project has it's own mechanics.. but this is likely the fifth time I have to rebase this, everytime with additional overhead in order to fix conflicts. :(

**Comment by pruiz** [comment]
Rebased, and tested locally (w/o k8s), let's hope CI tests finish ok. :?

**Comment by joestringer** [comment]
Github infrastructure appears to be having some trouble at the moment, causing image CI builds to fail. I'll try closing/reopening the PR to re-kick it.

**Comment by joestringer** [comment]
@pruiz I understand, thanks for bearing with us in this process. I'll keep an eye out for CI but unless something broke during the recent rebase I expect this should be OK to merge.

**Comment by pruiz** [comment]
> @pruiz I understand, thanks for bearing with us in this process. I'll keep an eye out for CI but unless something broke during the recent rebase I expect this should be OK to merge.

understood.. I'll try to keep an eye. Thnks

**Comment by joestringer** [comment]
`k8s-1.23-kernel-5.4` workflow hit known flaky issue #21735, no need to block this PR on the failure.

**Comment by joestringer** [comment]
That was the only failure, reviews are in, this looks good to merge. Thanks for your contribution! :tada:

**Comment by pruiz** [comment]
Nice! many thnks Joe!

**Pull request merged**
